src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,161 @@
+                }
+                else
+                {
+                    result = null;
+                }
+                return result;
+            }
+
+            internal void Add(PendingBranch branch)
+            {
+                var label = branch.Label;
+                if (label is null)
+                {
+                    _unlabeledBranches.Add(branch);
+                }
+                else
+                {
+                    var branches = GetOrAddBranchesToLabel(label);
+                    branches.Add(branch);
+                }
+            }
+
+            internal void AddRange(PendingBranchesCollection collection)
+            {
+                _unlabeledBranches.AddRange(collection._unlabeledBranches);
+                if (collection._labeledBranches is { })
+                {
+                    foreach (var pair in collection._labeledBranches)
+                    {
+                        var branches = GetOrAddBranchesToLabel(pair.Key);
+                        branches.AddRange(pair.Value.Branches);
+                    }
+                }
+            }
+
+            private ArrayBuilder<PendingBranch> GetOrAddBranchesToLabel(LabelSymbol label)
+            {
+                if (_labeledBranches is null)
+                {
+                    _labeledBranches = PooledDictionary<LabelSymbol, (int Index, ArrayBuilder<PendingBranch> Branches)>.GetInstance();
+                }
+                if (_labeledBranches.TryGetValue(label, out var pair))
+                {
+                    return pair.Branches;
+                }
+                else
+                {
+                    var branches = ArrayBuilder<PendingBranch>.GetInstance();
+                    _labeledBranches.Add(label, (_nextLabelIndex++, branches));
+                    return branches;
+                }
+            }
+
+            internal IEnumerable<PendingBranch> AsEnumerable()
+            {
+                return _labeledBranches is null ?
+                    _unlabeledBranches :
+                    asEnumerableCore();
+
+                IEnumerable<PendingBranch> asEnumerableCore()
+                {
+                    // Short _labeledBranches by label index.
+                    var labeledBranches = ArrayBuilder<(int Index, ArrayBuilder<PendingBranch> Branches)>.GetInstance();
+                    labeledBranches.AddRange(_labeledBranches.Values);
+                    labeledBranches.Sort((x, y) => x.Index - y.Index);
+
+                    foreach (var branch in _unlabeledBranches)
+                    {
+                        yield return branch;
+                    }
+
+                    foreach (var pair in labeledBranches)
+                    {
+                        foreach (var branch in pair.Branches)
+                        {
+                            yield return branch;
+                        }
+                    }
+
+                    labeledBranches.Free();
```

Comment:
This won't be executed if the `IEnumerator<T>` is not exhausted. Consider a `try / finally` to ensure `Free` is called in all cases.

Reply:
Done, thanks.